BigInt datatype implementation

seanhouser opened the pull request:
See closed PR #31 for additional history

seanhouser committed:
Progress

- Progress on subtraction, additional cases implemented and tested. Need
relational operators to continue
- Began implementation of relation operations
- Completed ==(BigInt) and ==(long long) with test cases in main all
successful

seanhouser committed:
Revert "Progress"

This reverts commit d7a1842855a7915b5ed42376c78c412690fde528.

seanhouser committed:
New Pull Request

Completed and tested
- Utility
- Contructors
- IO operators
- Assignment operators
- Unary operators
- Arithmetic operatos (partial)
- Arithmetic-assignment operators (partial)
- Relational operatos (partial)

seanhouser committed:
Completed all Subtraction and Addition operations for all cases incl. assignment and unary ops

- Completed and tested missing -x - (-y) subtraction case.
- Completed and tested missing x + (-y) and -x + y addition cases.
- Arithmetic-assignment operation -= works for BigInt and long long
params
- --BigInt and BigInt-- decrements as expected, following C++ semantics
- Switched arithmetic-assignment operations back to void return types

INCOMPLETE
- Division arithmetic and arithmetic-assignment operation
- Modulo arithmetic and arithmetic-assignment operation

varunjm commented on DataStructures/BigInteger/BigInt.cpp:
```diff
@@ -1,14 +1,261 @@
 */
 
 bool is_valid_number(const std::string& num) {
     for (char digit : num)
```
Using char& instead will make the code slightly faster.

faheel commented on DataStructures/BigInteger/BigInt.cpp:
```diff
@@ -1,14 +1,261 @@
-            return false;
+        if (digit < '0' || digit > '9')
+			if (digit != '+' && digit != '-')
+				return false;
```
`is_valid_number()` checks whether the magnitude of the number (which is a string) is a valid number, i.e. it should only have characters from `0` to `9`. So remove the checks for `+` and `-` because it makes the function work incorrectly.

faheel commented on DataStructures/BigInteger/BigInt.hpp:
```diff
@@ -20,53 +20,53 @@ class BigInt {
+    bool operator==(const BigInt&);	
+    bool operator==(long long);		
+    bool operator!=(const BigInt&);	
+    bool operator!=(long long);		
```
Please remove all trailing white-space.

faheel reviewed (changes requested):
Please keep the code for relational operators as contributed by @arielyyz [here](https://github.com/faheel/Algos/blob/dev/DataStructures/BigInteger/BigInt.cpp#L133)

seanhouser committed:
All functionality implemented - Review requests resolved

- Replaced relational operations code at @faheel request
- Removed trailing whitespaces from .hpp file
- Corrected is_valid_number() issue where sign was being checked, but
should not be
- All division cases implemented and tests in main are successful
- Division-assignment implemented and tested successfully in main
- All modulo cases implemented and tests in main are successful
- Modulo-assignment implemented and tested successfully in main

faheel commented:
@seanhouser I have completed the implementation as a separate library: https://github.com/faheel/BigInt

It is divided into separate header files, has unit tests for each operator, and I have used more efficient algorithms for math operators (such as Karatsuba's algorithm for multiplication, long division method for division, etc.). Do check it out!

faheel commented:
Once BigInt v1 is released, I plan to add its single-include header file to this repo (just like Catch) and use it for implementing algorithms that require large integers (such as some cryptographic algorithms). 

I could have merged this into the `dev` branch but there are still a few errors (such as incorrect return types and parameter types).

So, I'm sorry but I'm closing this PR now. Thanks a lot for your time and contribution!
